[BUG] Can't allocate object on stack if class has const members (despite having cpp_locals=True)

<!--
**PLEASE READ THIS FIRST:**
- Do not use the bug and feature tracker for support requests. Use the `cython-users` mailing list instead.
- Did you search for similar issues already? Please do, it helps to save us precious time that we otherwise could not invest into development.
- Did you try the latest master branch or pre-release? It might already have what you want to report. Also see the [Changelog](https://github.com/cython/cython/blob/master/CHANGES.rst) regarding recent changes.
-->

**Describe the bug**
Allocating a class with const members on the stack produces a .cpp file that can't be compiled because an instance of the class would be assigned to a an optional.

**To Reproduce**
Code to reproduce the behaviour:
- myclass.h:  
  ```cpp
  class MyClass {
  public:
      const int a;
      MyClass(int a) : a(a) {};
  };
  ```
- my_module.pxd
  ```pxd 
  cdef extern from "myclass.h":
      cdef cppclass MyClass:
          MyClass(int a) except +
  ```
- test.pyx
     ```cython
    # cython: language_level=3
    # distutils: language = c++
    # cython: cpp_locals=True
    
    from .my_module cimport MyClass
    
    cdef a_function():
         cdef MyClass instance = MyClass(1);
    ```
Compile using:
```
cython test.pyx --cplus
gcc -c test.cpp -I<python include directory> -std=c++17
```
Error message:
```
test.cpp: In function ‘PyObject* __pyx_f_7package_4test_a_function(int)’:
test.cpp:1849:26: error: use of deleted function ‘std::optional<MyClass>& std::optional<MyClass>::operator=(std::optional<MyClass>&&)’
 1849 |     __pyx_t_1 = MyClass(1);
      |                          ^
In file included from test.cpp:1185:
/usr/include/c++/10/optional:659:11: note: ‘std::optional<MyClass>& std::optional<MyClass>::operator=(std::optional<MyClass>&&)’ is implicitly deleted because the default definition would be ill-formed:
  659 |     class optional
      |           ^~~~~~~~
...
```

**Expected behavior**
The code to be generated correctly. My naive approach would be to replace the assignment with `std::optional::emplace(...)`, but I'm not sure if this would break other things.

**Environment (please complete the following information):**
 - OS: Debian 11
 - Python version: 3.9.7
 - Cython version: 3.0.0a10

--- a/Cython/Utility/CppSupport.cpp
+++ b/Cython/Utility/CppSupport.cpp
@@ -84,15 +84,50 @@ auto __Pyx_pythran_to_python(T &&value) -> decltype(to_python(
 ////////////// OptionalLocals.proto ////////////////
 //@proto_block: utility_code_proto_before_types
 
+#include <utility>
 #if defined(CYTHON_USE_BOOST_OPTIONAL)
     // fallback mode - std::optional is preferred but this gives
     // people with a less up-to-date compiler a chance
     #include <boost/optional.hpp>
-    #define __Pyx_Optional_Type boost::optional
+    #define __Pyx_Optional_BaseType boost::optional
 #else
     #include <optional>
     // since std::optional is a C++17 features, a templated using declaration should be safe
     // (although it could be replaced with a define)
     template <typename T>
-    using __Pyx_Optional_Type = std::optional<T>;
+    using __Pyx_Optional_BaseType = std::optional<T>;
 #endif
+
+// This class reuses as much of the implementation of std::optional as possible.
+// The only place it differs significantly is the assignment operators, which use
+// "emplace" (thus requiring move/copy constructors, but not move/copy
+// assignment operators). This is preferred because it lets us work with assignable
+// types (for example those with const members)
+template <typename T>
+class __Pyx_Optional_Type : private __Pyx_Optional_BaseType<T> {
+public:
+    using __Pyx_Optional_BaseType<T>::__Pyx_Optional_BaseType;
+    using __Pyx_Optional_BaseType<T>::has_value;
+    using __Pyx_Optional_BaseType<T>::operator*;
+    using __Pyx_Optional_BaseType<T>::operator->;
+#if __cplusplus >= 201103L || (defined(_MSC_VER) && _MSC_VER >= 1600)
+    __Pyx_Optional_Type& operator=(const __Pyx_Optional_Type& rhs) {
+        this->emplace(*rhs);
+        return *this;
+    }
+    __Pyx_Optional_Type& operator=(__Pyx_Optional_Type&& rhs) {
+        this->emplace(std::move(*rhs));
+        return *this;
+    }
+    template <typename U=T>
+    __Pyx_Optional_Type& operator=(U&& rhs) {
+        this->emplace(std::forward<U>(rhs));
+        return *this;
+    }
+#else
+    // Note - the "cpp_locals" feature is designed to require C++14.
+    // This pre-c++11 fallback is largely untested, and definitely won't work
+    // in all the cases that the more modern version does
+    using __Pyx_Optional_BaseType<T>::operator=; // the chances are emplace can't work...
+#endif
+};

--- a/tests/run/cpp_locals_directive.pyx
+++ b/tests/run/cpp_locals_directive.pyx
@@ -19,13 +19,9 @@ cdef extern from *:
             C(C&& rhs) : x(rhs.x), print_destructor(rhs.print_destructor) {
                 rhs.print_destructor = false; // moved-from instances are deleted silently
             }
-            C& operator=(C&& rhs) {
-                x=rhs.x;
-                print_destructor=rhs.print_destructor;
-                rhs.print_destructor = false; // moved-from instances are deleted silently
-                return *this;
-            }
-            C(const C& rhs) = default;
+            // also test that we don't require the assignment operator
+            C& operator=(C&& rhs) = delete;
+            C(const C& rhs) = delete;
             C& operator=(const C& rhs) = default;
             ~C() {
                 if (print_destructor) print_C_destructor();

